Test/QuickCheck/Rose.hs

```diff
@@ -0,0 +1,97 @@
+-- | A Rose 'Tree' based combinators to write generators and shrinkers
+-- simultaenously.
+--
+-- >>> let reven = fmap (*2) rarbitrary
+-- >>> quickCheck $ rForAll reven $ \x -> x < 4
+-- *** Failed! Falsifiable (after 6 tests and 1 shrink):
+-- 4
+--
+-- See <http://hypothesis.works/articles/integrated-shrinking/>
+module Test.QuickCheck.Rose
+  (
+  -- * Generators
+    RGen (..)
+  , rarbitrary
+  , roneof
+  , roneof'
+  , rsized
+  , rscale
+  -- * Properties
+  , rForAll
+  )
+ where
+
+import Control.Applicative
+import Control.Monad (join)
+import Data.List (inits)
+import Data.Traversable (traverse)
+import Data.Tree (Tree (..))
+import Test.QuickCheck
+import Test.QuickCheck.Property (Property (..))
+
+-- | A generator for values of type @a@ with built-in shrinking.
+--
+-- /Note:/ you should use applicative combinators (or @ApplicativeDo@),
+-- as much as possible to get quality shrinking.
+newtype RGen a = RGen { runRGen :: Gen (Tree a) }
+
+instance Functor RGen where
+    fmap f (RGen g) = RGen (fmap (fmap f) g)
+
+instance Applicative RGen where
+    pure = RGen . pure . pure
+    RGen f <*> RGen x = RGen $ (<*>) <$> f <*> x
+
+instance Monad RGen where
+    return = pure
+    RGen g >>= f = RGen $ do
+        tree  <- g
+        tree' <- traverse (runRGen . f) tree
+        pure $ join tree'
```

Comment:
I don't think this is the bind implementation you want.

`traverse` will split the random number generator for every node in `tree`, and this will cause unpredictable shrinking, because every time we shrink the `tree` it causes the rest of the computation `f` to be executed with a wildly different seed.

See the PR on `disorder-jack` where I fixed this same problem: https://github.com/ambiata/disorder.hs/pull/99